### PR TITLE
client authorization

### PR DESCRIPTION
**Pull request by billfort**
Signed-off-by: billfort <fxbao@hotmail.com>

- when a client connect to the node by websocket, the node will generate a challenge and send to the client
- when the client get the challenge, it sign the challenge by its private key, and send <i>setClient</i> action to the node, node will check if it has field <i>Signature</i>
- if have signature field, the node verify the signature by client public key.
- if pass verification, keep the connection, else disconnect it.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -428,6 +440,19 @@ func (c *Client) handleMessage(msgType int, data []byte) error {
...
+			var challenge string
+			if err := json.Unmarshal(*msg["Challenge"], &challenge); err != nil {
+				return err
+			}
+			byteChallenge, _ := hex.DecodeString(challenge)
+			signature, err := crypto.Sign(c.account.PrivateKey, byteChallenge)
+			if err != nil {
+				return err
+			}
+			c.chChallengeSignature <- signature
```
Comment (yilunzhang) [inline client.go:455]: Need to handle channel full (node somehow sent multiple challenge) and avoid blocking

File: client.go [hunk 2]
```diff
@@ -428,6 +440,19 @@ func (c *Client) handleMessage(msgType int, data []byte) error {
...
 			c.sigChainBlockHash = sigChainBlockHash
+
+		// client auth
+		case authChallengeAction:
+			var challenge string
+			if err := json.Unmarshal(*msg["Challenge"], &challenge); err != nil {
+				return err
+			}
+			byteChallenge, _ := hex.DecodeString(challenge)
+			signature, err := crypto.Sign(c.account.PrivateKey, byteChallenge)
```
Comment (yilunzhang) [inline client.go:451]: It is NOT safe to sign server provided data as it is. You need to make sure `byteChallenge` is NEVER a valid transaction digest.

File: client.go [hunk 3]
```diff
@@ -618,11 +643,25 @@ func (c *Client) connectToNode(node *Node) error {
...
+		case <-timer.C:
+		case signature = <-c.chChallengeSignature:
+		}
+
 		req := make(map[string]interface{})
 		req["Action"] = "setClient"
 		req["Addr"] = c.Address()
 
+		// client auth
+		if signature != nil {
```
Comment (yilunzhang) [inline client.go:661]: You can use `len(signature) > 0`

File: client.go (hunk 1 above)
Comment (billfort) [reply client.go:455]: fixed

File: client.go (hunk 2 above)
Comment (billfort) [reply client.go:463]: fixed. Add client salt to node challenge and sign on it.

File: client.go [hunk 4]
```diff
@@ -428,6 +446,31 @@ func (c *Client) handleMessage(msgType int, data []byte) error {
...
 			}
 			c.sigChainBlockHash = sigChainBlockHash
+
+		// client auth
+		case authChallengeAction:
+			var challenge string
+			if err := json.Unmarshal(*msg["Challenge"], &challenge); err != nil {
+				return err
+			}
+			byteChallenge, _ := hex.DecodeString(challenge)
```
Comment (yilunzhang) [inline client.go:456]: Handle error

File: client.go [hunk 5]
```diff
@@ -428,6 +446,31 @@ func (c *Client) handleMessage(msgType int, data []byte) error {
...
+
+			var clientSalt [32]byte
+			rand.Reader.Read(clientSalt[:])
```
Comment (yilunzhang) [inline client.go:459]: capture and handle error

File: client.go [hunk 6]
```diff
@@ -428,6 +446,31 @@ func (c *Client) handleMessage(msgType int, data []byte) error {
...
+
+			// add client salt to the node's challenge
+			byteChallenge = append(byteChallenge, clientSalt[:]...)
```
Comment (yilunzhang) [inline client.go:462]: I would suggest we sign `sha256(byteChallenge)`

File: client.go (hunk 4 above)
Comment (billfort) [reply client.go:456]: fixed

File: client.go (hunk 5 above)
Comment (billfort) [reply client.go:459]: fixed

File: client.go (hunk 6 above)
Comment (billfort) [reply client.go:462]: yes, fixed

**Comment by Phillipmmm** [comment]
I'm tying to get this going

File: client.go [hunk 7]
```diff
@@ -428,6 +446,38 @@ func (c *Client) handleMessage(msgType int, data []byte) error {
...
+			var challenge string
+			if err := json.Unmarshal(*msg["Challenge"], &challenge); err != nil {
+				return err
+			}
+			byteChallenge, err := hex.DecodeString(challenge)
+			if err != nil {
+				return err
+			}
+
+			clientSalt := make([]byte, 32)
```
Comment (yilunzhang) [inline client.go:461]: There is an existing helper function `RandomBytes()`

File: client.go [hunk 8]
```diff
@@ -428,6 +446,38 @@ func (c *Client) handleMessage(msgType int, data []byte) error {
...
+
+			// add client salt to the node's challenge
+			byteChallenge = append(byteChallenge, clientSalt...)
+			hash := sha256.Sum256(byteChallenge)
+			signature, err := crypto.Sign(c.account.PrivateKey, hash[:])
+			if err != nil {
+				return err
+			}
+
+			saltAndSig := stSaltAndSignature{ClientSalt: clientSalt[:], Signature: signature}
```
Comment (yilunzhang) [inline client.go:475]: `clientSalt[:]` is the same as `clientSalt`

File: client.go [hunk 9]
```diff
@@ -75,6 +80,15 @@ type Client struct {
 	node       *Node
 	sharedKeys map[string]*[sharedKeySize]byte
 	wallet     *Wallet
+
+	// client auth
+	chChallengeSignature chan stSaltAndSignature
```
Comment (yilunzhang) [inline client.go:85]: In general I would suggest we use pointer for better performance (this particular case doesn't have performance issue though)

File: client.go [hunk 10]
```diff
@@ -618,11 +668,26 @@ func (c *Client) connectToNode(node *Node) error {
...
+		case <-timer.C:
+		case saltAndSig = <-c.chChallengeSignature:
+		}
+
 		req := make(map[string]interface{})
 		req["Action"] = "setClient"
 		req["Addr"] = c.Address()
 
+		// client auth
+		if len(saltAndSig.Signature) > 0 {
```
Comment (yilunzhang) [inline client.go:686]: If we change to pointer, this needs to be `!= nil`

**Review by Phillipmmm — APPROVED** [review]

File: client.go (hunk 7 above)
Comment (billfort) [reply client.go:461]: Yes, modified.

File: client.go (hunk 8 above)
Comment (billfort) [reply client.go:475]: Yes, modified.

File: client.go (hunk 9 above)
Comment (billfort) [reply client.go:85]: Yes, modifed to pointer.

File: client.go (hunk 10 above)
Comment (billfort) [reply client.go:686]: Yes, modified.

**Review by yilunzhang — APPROVED** [review]

**Pull request merged**